src/index.d.ts

```diff
@@ -0,0 +1,344 @@
+     * via FarceActions.go
+     */
+    action: 'PUSH' | 'REPLACE' | 'POP'
+    /**
+     * the difference between the current index and the index of the previous location
+     */
+    delta: number
+    /**
+     * the location hash; as on window.location e.g. '#qux'
+     */
+    hash: string
+    /**
+     * the current index of the history entry, starting at 0 for the initial entry; this increments
+     * on FarceActions.push but not on FarceActions.replace
+     */
+    index: number
+    /**
+     * if present, a unique key identifying the current history entry
+     */
+    key?: string
+    /**
+     * the path name; as on window.location e.g. '/foo'
+     */
+    pathname: string
+    /**
+     * map version of search string
+     */
+    query: Map
+    /**
+     * the search string; as on window.location e.g. '?bar=baz'
+     */
+    search: string
+    /**
+     * additional location state that is not part of the URL
+     */
+    state: any
+  }
+
+  /**
+   * The shape might be different with a custom matcher or history enhancer, but the default matcher
+   * assumes and provides this shape. As such, this validator is purely for user convenience and
+   * should not be used internally.
+   */
+  interface Match {
+    location: Location
+    params: Map
+  }
+
+  /**
+   * An object implementing the matching algorithm.
+   *
+   * User code generally shouldn't need this, but it doesn't hurt to here, since we use it
+   * for routerShape below.
+   */
+  interface Matcher {
+    match: Function
+    getRoutes: Function
+    isActive: Function
+    /**
+     * Returns the path string for a pattern of the same format as a route path and a object of the
+     * corresponding path parameters
+     */
+    format: (pattern: any, params: Map) => any
+  }
+
+  /**
+   * Lenient arg version of location using in #push and #replace.
+   */
+  type LocationArg = Pick<Location, 'pathname'> &
+    Partial<Pick<Location, 'search' | 'hash' | 'state' | 'query'>>
+
+  /**
+   * The transition hook function receives the location to which the user is attempting to navigate.
+   *
+   * This function may return:
+   *  - true to allow the transition
+   *  - false to block the transition
+   *  - A string to prompt the user with that string as the message
+   *  - A nully value to call the next transition hook and use its return value, if present, or
+   *    else to allow the transition
+   *  - A promise that resolves to any of the above values, to allow or block the transition
+   *    once the promise resolves
+   *
+   * @see https://github.com/4Catalyzer/farce#transition-hooks
+   */
+  type TransitionHook = (
+    location: Location,
+  ) => undefined | (boolean | string | Promise<boolean | string>)
+
+  class Router {
+    /**
+     * Navigates to a new location
+     * @see farce
+     */
+    public push: (location: string | LocationArg) => void
```

Comment:
this actually does return whatever the action dispatch returns, but i'm guessing you wouldn't want to use that return value